src/GamePlayer.cpp

```diff
@@ -826,45 +826,50 @@ void GamePlayer::FindWarehouseForAllJobs(const Job job)
-    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true);
-
-    if(wh)
+    if(emergency)
     {
-        // Prüfe ob Notfallprogramm aktiv
-        if(!emergency)
-            return wh->OrderWare(ware, goal);
-        else
-        {
-            // Wenn Notfallprogramm aktiv nur an Holzfäller und Sägewerke Bretter/Steine liefern
-            if((ware != GD_BOARDS && ware != GD_STONES) || goal->GetBuildingType() == BLD_WOODCUTTER
-               || goal->GetBuildingType() == BLD_SAWMILL)
-                return wh->OrderWare(ware, goal);
-            else
-                return NULL;
-        }
-    } else // no warehouse can deliver the ware -> check all our wares for lost wares that might match the order
+        if(ware == GD_STONES && goal->GetBuildingType() != BLD_SAWMILL)
+            return NULL;
+
+        if(ware == GD_BOARDS && (goal->GetBuildingType() != BLD_WOODCUTTER && goal->GetBuildingType() != BLD_SAWMILL))
+            return NULL;
+    }
+
+    // Look for available ware in a warehouse
+    unsigned int nearestWarehouse = std::numeric_limits<unsigned>::max();
+    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true, &nearestWarehouse);
+
+    // Look for ware waiting on the road or on the way to storehouse (if addon is enabled)
+    unsigned nearestWare = std::numeric_limits<unsigned>::max();
+    Ware* bestWare = NULL;
+    const bool improvedWareOrder = gwg->GetGGS().isEnabled(AddonId::IMPROVED_WARE_ORDER);
+    if(!wh || improvedWareOrder)
     {
-        unsigned bestLength = std::numeric_limits<unsigned>::max();
-        Ware* bestWare = NULL;
         BOOST_FOREACH(Ware* curWare, ware_list)
         {
-            if(curWare->IsLostWare() && curWare->type == ware)
+            noBaseBuilding* wareGoal = curWare->GetGoal();
+            // Ware is redirectable if its lost (has no goal) or its goal is a storehouse
+            if(curWare->type == ware && (curWare->IsLostWare() || (improvedWareOrder && wareGoal->GetBuildingType() == BLD_STOREHOUSE)))
             {
-                // got a lost ware with a road to goal -> find best
                 unsigned curLength = curWare->CheckNewGoalForLostWare(goal);
-                if(curLength < bestLength)
+                if(curLength < nearestWare)
                 {
-                    bestLength = curLength;
+                    nearestWare = curLength;
                     bestWare = curWare;
                 }
             }
         }
-        if(bestWare)
-        {
-            bestWare->SetNewGoalForLostWare(goal);
-            return bestWare;
-        }
+    }
+
+    // order ware in Warehouse if is near then available ware
+    if(wh && nearestWarehouse < nearestWare)
+    {
+        return wh->OrderWare(ware, goal);
+    } else if(bestWare) // any ware was found
+    {
+        // Change Goal for Ware from NULL/Warehouse to orderer.
+        bestWare->SetNewGoalForLostWare(goal);
```

Comment:
When the ware had a goal this might be a problem. First the warehouse has to be notified that the ware won't arrive. 2nd: What about cases, where the warehouse (e.g. harbour) ordered the ware? 3rd: that case might even lead to a bigger problem: When cancelling the ware, the building might reorder it immediately (in the cancel method) and could find the exact same ware. Then it orders that ware again but you overwrite its goal to something else.

I think I know now why this hasn't been implemented earlier ;)

Reply:
Eine kleinere Frage hier auf deutsch, ich denke, dass es gerade einfacher ist für mich dies auf deutsch zu beschreiben.

Dass ich das Lagerhaus/HQ/Hafen benachrichtigten muss, verstehe ich; irgendwie habe ich ignoriert, dass der Bestellende noch auf seine Ware wartet. 
Die Kritik, dass ein CancelWare ggf. zu einer Neubestellung führen könnte verstehe ich auch, da die Ware bestellbar scheint, wenn sie kein Ziel hat (somit als "verlorene" Ware gilt).
 
Müsste es aber nicht "ausreichen", wenn ich das Ziel der Ware ändere, dem neuen Ziel zuordne  (TakeWare) und dann auf dem alten Ziel CancelWare aufrufe? Hierdurch müsste ich eigentlich sicherstellen, dass die Ware bereits "benutzt" wird, bevor das Lagerhaus sie bestellen könnte.